Fix initial optimizer settings in external tests

cameel opened the pull request:
Currently external test scripts rely on the `$OPTIMIZER_LEVEL` variable to set minimal options that the external project works with. This variable is only used just before running tests though. If the project runs contract compilation during JS package installation (as it happens in case of Gnosis, see https://github.com/ethereum/solidity/pull/10336#issuecomment-740106553), it's still using initial settings that do not set optimization level at all. This PR fixes fixes that by making the initial config use the lowest allowed optimization level.

The core of this PR is this change:
```patch
-module.exports['compilers'] = {solc: {version: "$dir/solc"} };
+module.exports['compilers'] = {solc: {version: "$dir/solc", settings: $(solc_settings "$OPTIMIZER_LEVEL" istanbul)}};
```
the rest is just a refactor to make that change easy without code duplication.

cameel committed:
Make printError() print to stderr

cameel committed:
Refactor truffle_run_test() to make optimizer settings reusable

cameel committed:
Take into account optimization level when creating initial Truffle settings for external tests

bshastry commented on test/externalTests/common.sh:
```diff
@@ -153,27 +153,25 @@ function force_solc
 {
     local config_file="$1"
     local settings="$2"
-    local evmVersion="$3"
```
Note to self: this function now accepts two instead of three arguments

cameel replied:
And back to four :P
I wouldn't get too attached to the current signatures of these functions :)

bshastry commented on test/externalTests/common.sh:
```diff
@@ -153,27 +153,25 @@ function force_solc
     printLog "Forcing solc version..."
     cat >> "$config_file" <<EOF
-module.exports['compilers'] = {solc: {version: "$dir/solc"} };
+module.exports['compilers'] = {solc: {version: "$dir/solc", settings: $(solc_settings "$OPTIMIZER_LEVEL" istanbul)}};
```
This is a bit confusing at least for me because function is called `force_solc` but it is also forcing a setting. There is another function called `force_solc_settings` that is also doing the same?

cameel replied:
Yeah, that part is confusing. I didn't want to refactor too much in this PR because it's mostly meant as fix but since you pointed it out, I added a few more commits that clean this up.

I have renamed these functions and removed `force_solc` so that now you need to separately change the config and overwrite `solc` npm module. The only important functional change is the fact that now there's no special initial config - instead it always specifies all the options (i.e. both the compiler path/version and the optimizer settings). I put that one in a commit separate from others that are pure refactors plus very minor functional changes (slightly different messages printed to stdout and the fact that `truffle_run_test` no longer sets the initial config just before immediately overwriting it).

bshastry commented on test/externalTests/common.sh:
```diff
@@ -215,6 +213,27 @@ function run_test
+            exit 1
+            ;;
+    esac
+}
```
Nice, this refactor made the code more readable :-)

bshastry reviewed (changes requested):
Minor comment regarding the use of `force_solc` and `force_solc_settings` functions

cameel committed:
force_solc_settings: Stop using the outdated 'solc' key for compiler settings in truffle-config.js

- Settings were moved under 'compilers' in Truffle 5.0.0 and none of the projects use older Truffle.

cameel committed:
Remove unused find_truffle_config

cameel committed:
When changing compiler settings in Truffle config, always set all of them

cameel committed:
Simplify functions for setting solc version in Truffle config

bshastry reviewed (approved):
LGTM, thanks for the clarifications!
